### PR TITLE
enh(monitoring): mechanism to switch repository

### PR DESCRIPTION
**Pull request by kduret**
## Description

This PR intends to add a new option to Administration -> Parameters -> Monitoring page to allow the switch repositories for Resource Status page.

image

**Fixes** MON-13318

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 20.10.x
- [ ] 21.04.x
- [ ] 21.10.x
- [x] 22.04.x (master)

### CODE REVIEW

File: www/include/Administration/parameters/DB-Func.php [hunk 1]
```diff
@@ -690,6 +725,21 @@ function updateGeneralConfigData($gopt_id = null)
...
+        : 'legacy';
+    $dotenvVars = var_export($dotenvVars, true);
+    $dotenvVars = <<<EOF
+    <?php
+
+    return $dotenvVars;
+
+    EOF;
+
+    file_put_contents($dotenvPath, $dotenvVars, \LOCK_EX);
```
Comment (kduret) [inline www/include/Administration/parameters/DB-Func.php:741]: create a separated function to avoid confusion with fonction name ...InDb

Comment (kduret) [inline www/include/Administration/parameters/DB-Func.php:741]: check file_put_contents does not return false (throws exception otherwise)

Comment (kduret) [reply www/include/Administration/parameters/DB-Func.php:741]: if throwing exception print blank screen, then just log error message

File: www/include/Administration/parameters/DB-Func.php [hunk 2]
```diff
@@ -693,6 +728,43 @@ function updateGeneralConfigData($gopt_id = null)
...
+/**
+ * Update Resource Status environment variable in PHP file.
+ *
+ * @throws \Exception
+ */
+function updateResourceStatusMode(): void
+{
+    global $form;
+
+    $ret = array();
```
Comment (kduret) [inline www/include/Administration/parameters/DB-Func.php:740]:
```suggestion
```

File: config/packages/Centreon.yaml [hunk 3]
```diff
@@ -244,19 +243,23 @@ services:
     Core\Infrastructure\Security\Api\FindProviderConfigurations\FindProviderConfigurationsPresenter:
         arguments: [!tagged_iterator 'authentication.provider.presenters']
 
-    Centreon\Domain\Monitoring\Interfaces\ResourceRepositoryInterface:
-        class: Centreon\Infrastructure\Monitoring\Resource\ResourceRepositoryRDB
+    Centreon\Domain\Monitoring\Interfaces\ResourceServiceInterface:
+        class: Centreon\Domain\Monitoring\ResourceService
+        calls:
+            - method: setResourceRepository
+              arguments: ['@=parameter("resource.status.repository") === "legacy" ? service("Centreon\\Infrastructure\\Monitoring\\Resource\\ResourceRepositoryRDB") : service("Centreon\\Infrastructure\\Monitoring\\Resource\\DbReadResourceRepository")']
```
Comment (jeremyjaouen) [inline config/packages/Centreon.yaml:250]: You should use a constant for "legacy"

File: www/include/Administration/parameters/DB-Func.php [hunk 4]
```diff
@@ -147,6 +147,41 @@ function isSessionDurationValid(int $value = null)
...
+ * @param int $value
+ * @return bool
+ */
+function canUseResourceStatusOptimized($value)
+{
+    global $pearDB;
+
+    $mode = $value['resource_status_mode'];
+
+    if ($mode === 'legacy') {
```
Comment (jeremyjaouen) [inline www/include/Administration/parameters/DB-Func.php:162]: You should use a constant here.

File: www/include/Administration/parameters/DB-Func.php [hunk 5]
```diff
@@ -693,6 +728,41 @@ function updateGeneralConfigData($gopt_id = null)
     $centreon->initOptGen($pearDB);
 }
 
+/**
+ * Update Resource Status environment variable in PHP file.
+ *
+ * @throws \Exception
+ */
+function updateResourceStatusMode(): void
```
Comment (jeremyjaouen) [inline www/include/Administration/parameters/DB-Func.php:736]: I'm not sure this is the best place for this function, IMO this file should only contains operation on database. and switching an environment variable is not appropriate here (maybe better in www/include/common/common-Func.php)

File: www/include/Administration/parameters/general/form.php [hunk 6]
```diff
@@ -32,7 +33,11 @@
  * For more information : contact@centreon.com
  *
  */
-require_once _CENTREON_PATH_ . "www/class/centreonGMT.class.php";
+
+require_once __DIR__ . '/../../../../../bootstrap.php';
+require_once __DIR__ . "/../../../../class/centreonGMT.class.php";
+
+use Symfony\Component\Dotenv\Dotenv;
```
Comment (jeremyjaouen) [inline www/include/Administration/parameters/general/form.php:40]: if you use Dotenv you can retrieve your file as follow:

$dotenv = new Dotenv();
$dotenv->load(__DIR__.'/.env');

then access your .env variables as $_ENV['MY_ENV_VARIABLE_NAME']

https://symfony.com/doc/4.1/components/dotenv.html

Comment (kduret) [reply www/include/Administration/parameters/general/form.php:40]: This line should be removed
I think it's better to get parameter from container, cause if the value is not defined in environment, it uses the default value set in the config

**Review by jeremyjaouen — DISMISSED** [review]

File: www/include/Administration/parameters/general/help.php [hunk 7]
```diff
@@ -69,6 +69,14 @@
 $help['tip_sort_problems_by'] = dgettext('help', 'Default sort in monitoring consoles.');
 $help['tip_order_sort_problems'] = dgettext('help', 'Default order in monitoring consoles.');
 
+/**
+ * Monitoring
+ */
+$help['tip_resource_status_mode'] = dgettext(
+    'help',
+    'Mode to get monitoring information in resource status page.'
```
Comment (a-launois) [inline www/include/Administration/parameters/general/help.php:77]: There is no traduction for this ?

Comment (adr-mo) [reply www/include/Administration/parameters/general/help.php:77]: I dunno if help tooltips are translatable, I'll check and do the translation if needed 👍

Comment (adr-mo) [reply www/include/Administration/parameters/general/help.php:77]: @a-launois done :)

**Review by jeremyjaouen — DISMISSED** [review]

File: src/Centreon/Infrastructure/Monitoring/Resource/ResourceRepositoryFactory.php [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    private static ?ResourceRepositoryInterface $repositoryToFind = null;
+
+    /**
+     * @param ResourceRepositoryInterface $repositoryToFind
+     */
+    public function __construct(ResourceRepositoryInterface $repositoryToFind)
+    {
+        self::$repositoryToFind = $repositoryToFind;
+    }
+
```
Comment (adr-mo) [inline src/Centreon/Infrastructure/Monitoring/Resource/ResourceRepositoryFactory.php:41]: Add doc bloc

**Review by TamazC — DISMISSED** [review]

File: www/include/Administration/parameters/general/help.php (hunk 7 above)
Comment (adr-mo) [inline www/include/Administration/parameters/general/help.php:77]: Explain that this will only work with BBDO 3.0.0

**Review by callapa — DISMISSED** [review]

File: lang/fr_FR.UTF-8/LC_MESSAGES/help.po [hunk 9]
```diff
@@ -7214,3 +7214,8 @@ msgstr ""
 
 msgid "Expected TLS certificate common name (CN) - leave blank if unsure."
 msgstr "Champ common name (CN) attendu dans le certificat TLS - laisser vide en cas de doute"
+
+msgid "Mod used to retrieve information displayed in Resource Status page."
+"The optimized mod will only be available and functional when using the BBDO 3.0.0 protocol for the collect engine."
```
Comment (a-launois) [inline lang/fr_FR.UTF-8/LC_MESSAGES/help.po:7219]:
```suggestion
msgid "Mode used to retrieve information displayed in Resource Status page."
"The optimized mode will only be available and functional when using the BBDO 3.0.0 protocol for the collect engine."
```

File: www/include/Administration/parameters/general/help.php [hunk 10]
```diff
@@ -69,6 +69,15 @@
 $help['tip_sort_problems_by'] = dgettext('help', 'Default sort in monitoring consoles.');
 $help['tip_order_sort_problems'] = dgettext('help', 'Default order in monitoring consoles.');
 
+/**
+ * Monitoring
+ */
+$help['tip_resource_status_mode'] = dgettext(
+    'help',
+    'Mod used to retrieve information displayed in Resource Status page. The optimized mod will only be available '
```
Comment (a-launois) [inline www/include/Administration/parameters/general/help.php:77]:
```suggestion
    'Mode used to retrieve information displayed in Resource Status page. The optimized mode will only be available '
```

**Review by jeremyjaouen — DISMISSED** [review]

**Review by a-launois — DISMISSED** [review]

**Review by callapa — DISMISSED** [review]

**Review by a-launois — DISMISSED** [review]

**Review by callapa — DISMISSED** [review]

**Review by jeremyjaouen — DISMISSED** [review]

**Review by jeremyjaouen — DISMISSED** [review]

File: src/Centreon/Infrastructure/Monitoring/Resource/ResourceRepositoryFactory.php [hunk 11]
```diff
@@ -0,0 +1,49 @@
+<?php
+
+/*
+ * Copyright 2005 - 2021 Centreon (https://www.centreon.com/)
```
Comment (adr-mo) [inline src/Centreon/Infrastructure/Monitoring/Resource/ResourceRepositoryFactory.php:4]:
```suggestion
 * Copyright 2005 - 2022 Centreon (https://www.centreon.com/)
```

**Review by a-launois — APPROVED** [review]

**Review by TamazC — APPROVED** [review]

**Pull request merged**